include/WCSimAmBeGen.hh

```diff
@@ -0,0 +1,56 @@
+#ifndef WCSimAmBeGen_h
+#define WCSimAmBeGen_h
+
+#include "G4SPSEneDistribution.hh"
+#include "G4SPSRandomGenerator.hh"
+#include "G4RandomDirection.hh"
+
+#include "globals.hh"
+#include "jhfNtuple.h"
+#include <fstream>
+
+using namespace std;
+
+class G4ParticleGun;
+class G4Event;
+
+class WCSimAmBeGen
+{
+  public:
+    WCSimAmBeGen();
+    ~WCSimAmBeGen();
+
+    // Initialise the AmBe generator
+    void Initialise();
+    //void ReadFromFile();
+    
+    // Set Particle Gun Properties of the neutron and gamma
+    G4double NeutronEnergy(G4double gamEnergy);
+    G4double GammaEnergy();
+    void GenerateNG(G4Event* anEvent);
+
+  private:
+    G4ParticleGun*        myAmBeGun;
+    G4SPSRandomGenerator* rGen;
+    G4SPSEneDistribution* nEnergyDist;
+    
+    // Variables for the initialisation of AmBe generator parameters
+    G4double nEnergy;
```

Comment:
```suggestion
```
Only used inside `NeutronEnergy()`. Doesn't need to be a class variable

Reply:
<img width="555" alt="Captura de pantalla 2024-02-29 a las 16 39 42" src="https://github.com/WCSim/WCSim/assets/67594492/9379b885-7269-43c7-945e-168a533679fe">
<img width="547" alt="Captura de pantalla 2024-02-29 a las 16 39 49" src="https://github.com/WCSim/WCSim/assets/67594492/1add87ff-8012-47f8-a335-33386ed1c06a">

Reply:
Hmm that's weird. Can you commit & push so I can see the code?

Reply:
I've just checked & it's getting into the correct `if`s. So its probably not switching histofiles correctly

You may need to setup `nEnergyDist` differently, as it may not be switching between them correctly
So you'd want to 
- make 3 class variables (instead of just one) `nEnergyDistX` where `X` is some label for each
- in `Initialise()` do the stuff like this
```  
nEnergyDist->ArbEnergyHistoFile(gs_path);
nEnergyDist->SetBiasRndm(rGen);
nEnergyDist->ArbInterpolate("Lin");
```
- in the `NeutronEnergy()` `if`s, select the correct `nEnergyDistX`

This'll also be more efficient, as it won't be interpolating every event

Reply:
I tried that previously in order to make it more efficient, as you said, but I ended up returning to the "just one nEnergyDist" situation because of this error:
```
### Run 0 starts.

-------- EEEE ------- G4Exception-START -------- EEEE -------
*** G4Exception : Event0302
      issued by : G4SPSEneDistribution::ArbInterpolate
Error: this is for arbitrary distributions
*** Fatal Exception *** core dump ***
-------- EEEE -------- G4Exception-END --------- EEEE -------


*** G4Exception: Aborting execution ***
```
I changed the code so now it looks like this:

<img width="397" alt="Captura de pantalla 2024-02-29 a las 18 43 29" src="https://github.com/WCSim/WCSim/assets/67594492/1b5141f5-4ee7-4140-9114-57b34981a62b">

<img width="563" alt="Captura de pantalla 2024-02-29 a las 18 42 55" src="https://github.com/WCSim/WCSim/assets/67594492/dae1e7c9-480d-469d-941c-9e1a3aba3501">
492/d501d540-756d-43c8-b35b-aed237b57594">

Reply:
I think you need to do `SetEnergyDisType("Arb")` before `ArbInterpolate("Lin")`. This is what the error complains about

Reply:
I have also tried setting the histogram options first and then the bias and it does not work either.

Reply:
In your snippet above, `rGen` is set after it is used in the `nEnergyDist*` setups. This could be the issue (if it is, the error isn't very clear!)

Reply:
Changed it to this, still same error 😞 
```
void WCSimAmBeGen::Initialise(){
    rGen          = new G4SPSRandomGenerator();
    myAmBeGun     = new G4ParticleGun();

    nEnergyDistGS = new G4SPSEneDistribution();
    nEnergyDistGS->SetEnergyDisType("Arb");
    nEnergyDistGS->ArbEnergyHistoFile(gs_path);
    nEnergyDistGS->ArbInterpolate("Lin");
    nEnergyDistGS->SetBiasRndm(rGen);

    nEnergyDistFE = new G4SPSEneDistribution();
    nEnergyDistFE->SetEnergyDisType("Arb");
    nEnergyDistFE->ArbEnergyHistoFile(fe_path);
    nEnergyDistFE->ArbInterpolate("Lin");
    nEnergyDistFE->SetBiasRndm(rGen);

    nEnergyDistSE = new G4SPSEneDistribution();
    nEnergyDistSE->SetEnergyDisType("Arb");
    nEnergyDistSE->ArbEnergyHistoFile(se_path);
    nEnergyDistSE->ArbInterpolate("Lin");
    nEnergyDistSE->SetBiasRndm(rGen);
   
    vtx = G4ThreeVector(0., 0., 0.);
    time = 0.;
    epsilon = 1e-6;
}

```

Reply:
This can be resolved, once you remove the `G4double nEnergy;` in the header file (currently a compilation warning)

Reply:
Done 👍